text/0002-buildpack-toml-in-app.md

```diff
@@ -0,0 +1,187 @@
+uri = "<url or path to the buildpack (optional default=urn:buildpack:<id>)"
+```
+
+* `id` (string, required): the ID ([as defined in the spec](https://github.com/buildpack/spec/blob/master/buildpack.md#buildpacktoml-toml) of another buildpack
+* `version` (string, optional, default=`latest`): the version ([as defined in the spec](https://github.com/buildpack/spec/blob/master/buildpack.md#buildpacktoml-toml) of the dependent buildpack. The default is to use the latest available version of the buildpack (resolution of this value may be platform-dependent).
+* `optional` (bool, optional, default=`false`): Defines whether this buildpack will be optional in `group.toml`.
+* `uri` (string, optional, default=`urn:buildpack:<id>`): The exact location of the dependent buildpack. If not specified the platform will resolve the `urn:buildpack:<id>` (making the resolution platform dependent).
+
+## New Element: project.env
+
+The `project.env` element may contain a list of key value pairs:
+
+```toml
+[project.env]
+KEY = "<value>"
+```
+
+The key-value pairs will be set as environment variables during buildpack execution on this repo.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Given the elements described above, we seek to satisfy the following use cases:
+
+## Inline buildpacks
+
+When a `buildpack.toml` is found in the root directory of an app, and it contains the following element:
+
+```toml
+[buildpack]
+id = "<buildpack id>"
+```
+
+Then the lifecycle will add the app repo (as a buildpack) to the end of every group in the pre-defined buildpack groups.
+
+Optionally, the `buildpack.toml` may include:
+
+```toml
+[buildpack]
+root = "<directory name>" # the directory where the `bin/` scripts are located
+```
+
+The `root` element is necessary to support apps that may contain a buildpack in a subdirectory.
+
+## Monorepos
+
+When a `buildpack.toml` is found in an app directory, the lifecycle will check for the following element:
+
+```toml
+[project]
+root = "<directory name>" # must be a relative path and a subdir. no `..`
+```
+
+If `project.root` is set, the lifecycle will execute against this directory instead of the directory that contains the `buildpack.toml`. All other configuration in the `buildpack.toml` will still be honored, and a `buildpack.toml` in the new `root` will be ignored (we may want to reconsider that however).
+
+## Codified Buildpacks
+
+Given an app with a `buildpack.toml`, the lifecycle will read the `project.buildpacks` and construct an `group.toml` that overrides the default buildpack groups in `order.toml`. Only the buildpacks listed in `project.buildpacks` will be run. For example, an app might contain:
+
+```toml
+[[project.buildpacks]]
+id = "heroku/java"
+```
+
+If the app also contains a `Gemfile` and the `heroku/buildpacks` builder image is used, this will override the default buildpack groups, which would detect and run the `heroku/ruby` buildpack  by default.
+
+## Buildpack Dependencies
```

Comment:
Some thoughts about this:
1. It doesn't allow the user to run one group after another group without manually specifying each buildpack. I'd like the ability to name an `order.toml`-like definition and reference it inside other `order.toml`-like definitions. I feel like it's important to abstract away the modularity for end users.
2. I feel weird about encoding the order definition in each modular buildpack, as opposed to defining that order definition separately (but maybe giving it an ID that's interchangable with a buildpack ID, as in #2).

Reply:
Originally I was hoping we could simplify it so that only one group could be specified. But I'm starting to think that was misguided. More to come...